### PR TITLE
[do-not-merge] Replace CircleCI with GitHub Actions

### PR DESCRIPTION
**Pull request by drpaneas**
This PR is creating the following Makefile:

```shell
$ make help    
Usage:
  all     Default target, now is 'build'
  build   Builds the binary
  setup   Runs mod download and generate
  test    Runs the tests with coverage
  cover   Runs all tests and opens the coverage report in the browser
  lint    Runs golangci-lint (configuration at .golangci.yml) and misspell
  clean   Cleans the binaries and the Go module cache
  run     Runs go run
  race    headless test with xvfb (runs only on Linux)
  help    Prints this help message
```

All of those (apart from `race`) are working for all systems (Linux, MacOS, Windows). The only requirement for the user/dev is to have Golang installed (which they can do by running the `build.sh` we already have, if they don't have it installed) and the ability to run `make`.

The GitHub action 'Build" is creating 3 jobs, one per operating system. All of the steps (apart from the Go installation) is driven by the makefile. As a result, a developer would be 100% able to replicate locally what the CI is running. That said, it's encourage to run that stuff locally before submitting your PR, so you can quickly see what's going to happen on the CI without surprises.

The CI is running the following order:

```
# git clone the project
# install Go
make setup
make lint
make test
make race
make build
```

In case you are wondering, the `golangci-lint` is installed via the Go modules (and gets updated when you run `go get -u` which is already part of the makefile) when the user is running `make setup` and is running as `make lint`. Tests are running by `make test` and the `/usr/bin/xvfb-run --auto-servernum go test -v -race ./...` is running via `make race` and it's triggered only for Linux.

last bot not least, obviously this PR is removing CircleCI.
You can see a successful run: https://github.com/drpaneas/HellSpawner/runs/2223612441?check_suite_focus=true

### CODE REVIEW

**Comment by drpaneas** [comment]
I removed the `.circleci` folder but I guess this was not enough. Please admins take a look at the GitHub settings for this to disable it.

File: Makefile [hunk 1]
```diff
@@ -24,10 +27,26 @@ build:
...
+ifdef REDHATOS
+	@echo "Downloading packages for RedHat based..."
+	@sudo dnf install -y xorg-x11-server-Xvfb libX11-devel libXcursor-devel libXrandr-devel libXinerama-devel mesa-libGL-devel alsa-lib-devel libXi-devel
+endif
+	@echo "Run: git submodule update --init --recursive"
 	@git submodule update --init --recursive
+	@echo "Run: go get -d"
 	@$(GOCMD) get -d
+	@echo "Run: go get -u"
+	@$(GOCMD) get -u
```
Comment (gucio321) [inline Makefile]: I don't think, that workflow (and anyone, who uses `make setup`) should update dependencies. It chould cause crasches.

Comment (drpaneas) [reply Makefile]: True, this is too intrusive. I will separate this in different `make` step.

File: go.mod [hunk 2]
```diff
@@ -6,23 +6,42 @@ require (
 	github.com/OpenDiablo2/OpenDiablo2 v0.0.0-20210329094534-c8ef8cf2a46d
 	github.com/OpenDiablo2/dialog v0.0.0-20201230220514-26162241209f
 	github.com/TheTitanrain/w32 v0.0.0-20200114052255-2654d97dbd3d // indirect
+	github.com/client9/misspell v0.3.4
 	github.com/enriquebris/goconcurrentqueue v0.6.0
 	github.com/faiface/beep v1.0.2
 	github.com/go-gl/gl v0.0.0-20210315015930-ae072cafe09d // indirect
 	github.com/go-gl/glfw/v3.3/glfw v0.0.0-20210311203641-62640a716d48
 	github.com/go-resty/resty/v2 v2.5.0 // indirect
+	github.com/golangci/golangci-lint v1.39.0
```
Comment (gucio321) [inline go.mod:15]: good catch!

**Review by gucio321 — APPROVED** [review]

**Comment by gravestench** [comment]
very cool

**Review by gravestench — APPROVED** [review]

**Comment by gucio321** [comment]
@drpaneas maybe we can use golangci-lint action: https://github.com/golangci/golangci-lint-action

**Comment by drpaneas** [comment]
> @drpaneas maybe we can use golangci-lint action: https://github.com/golangci/golangci-lint-action

We can use that sure, but it then we will have inconsistencies between the CI and the local setup. This is because the devs will have to individually take care of downloading golangci and keep the version in sync with the repo. Instead, the way I am using it at this PR, golangci is getting installed via to modules, so the dev doesn't have to care about it. The downside is that we have to keep updating it in the repo, while if we use the GH action it will use the latest by default.

I am fine with either way. Should I change the PR?

**Comment by drpaneas** [comment]
> @drpaneas maybe we can use golangci-lint action: https://github.com/golangci/golangci-lint-action

Actually I found this https://golangci-lint.run/usage/install/#install-from-source

which states that my workflow (using Go modules) to fetch it it's unstable. I guess moving it via the GitHub actions it's better then. I will work on it.

**Comment by drpaneas** [comment]
See: https://github.com/OpenDiablo2/HellSpawner/pull/273

**Pull request closed without merging**